sway/input/text_input.c

```diff
@@ -255,6 +267,240 @@ static void relay_handle_text_input(struct wl_listener *listener,
+		popup->scene_tree = NULL;
+		return;
+	}
+
+	if (!cursor_rect) {
+		cursor_area.x = 0;
+		cursor_area.y = 0;
+		cursor_area.width = parent.width;
+		cursor_area.height = parent.height;
+	}
+
+	int popup_width = popup->popup_surface->surface->current.width;
+	int popup_height = popup->popup_surface->surface->current.height;
+	int x1 = cursor_area.x;
+	int x2 = cursor_area.x + cursor_area.width;
+	int y1 = cursor_area.y;
+	int y2 = cursor_area.y + cursor_area.height;
+	int relate_x = x1;
+	int relate_y = y2;
+
+	int available_right = output_box.x + output_box.width - x1;
+	int available_left = x2 - output_box.x;
+	if (available_right < popup_width && available_left > available_right) {
+		relate_x = x2 - popup_width;
+	}
+
+	int available_down = output_box.y + output_box.height - y2;
+	int available_up = y1 - output_box.y;
+	if (available_down < popup_height && available_up > available_down) {
+		relate_y = y1 - popup_height;
+	}
+	int x = parent.x + relate_x;
+	int y = parent.y + relate_y;
+
+	wlr_scene_node_set_position(&relative->node, relate_x, relate_y);
+	if (cursor_rect) {
+		struct wlr_box box = {
+			.x = x1 - relate_x,
+			.y = y1 - relate_y,
+			.width = cursor_area.width,
+			.height = cursor_area.height,
+		};
+		wlr_input_popup_surface_v2_send_text_input_rectangle(
+			popup->popup_surface, &box);
+	}
+	wlr_scene_node_set_position(&popup->scene_tree->node, x, y);
+}
+
+static void input_popup_set_focus(struct sway_input_popup *popup,
+		struct wlr_surface *surface) {
+	wl_list_remove(&popup->focused_surface_unmap.link);
+
+	if (surface == NULL) {
+		wl_list_init(&popup->focused_surface_unmap.link);
+		input_popup_update(popup);
+		return;
+	}
+	struct wlr_layer_surface_v1 *layer_surface =
+		wlr_layer_surface_v1_try_from_wlr_surface(surface);
+	if (layer_surface != NULL) {
+		wl_signal_add(
+			&layer_surface->surface->events.unmap, &popup->focused_surface_unmap);
+		input_popup_update(popup);
+		return;
+	}
+
+	struct sway_view *view = view_from_wlr_surface(surface);
+	wl_signal_add(&view->events.unmap, &popup->focused_surface_unmap);
+}
+
+static void handle_im_popup_destroy(struct wl_listener *listener, void *data) {
+	struct sway_input_popup *popup =
+		wl_container_of(listener, popup, popup_destroy);
+	wl_list_remove(&popup->focused_surface_unmap.link);
+	wl_list_remove(&popup->popup_surface_commit.link);
+	wl_list_remove(&popup->popup_destroy.link);
+	wl_list_remove(&popup->link);
+	if (popup->scene_tree != NULL) {
+		wlr_scene_node_destroy(&popup->scene_tree->node);
+	}
+
+	free(popup);
+}
+
+static void handle_im_popup_surface_commit(struct wl_listener *listener,
+		void *data) {
+	struct sway_input_popup *popup =
+		wl_container_of(listener, popup, popup_surface_commit);
+	input_popup_update(popup);
+}
+
+static void handle_im_focused_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct sway_input_popup *popup =
+		wl_container_of(listener, popup, focused_surface_unmap);
+	input_popup_update(popup);
```

Comment:
I think we are leaking memory here. `input_popup_update` will return early due to the check for `->mapped` and not destroy the old scene nodes.

Reply:
but if not mapped, the destory will cause coredump. then how to solve it ?

Reply:
It might be that the next time that sway tries to map a input popup, it will double free the scene nodes.

Reply:
Try nulling your pointers as you destroy the scene.

Reply:
Emm. am I doing right?

Reply:
Hm, I don't think so. The pointers are NULL'ed but the scene nodes are not destroyed. How about the following:

```c
static void input_popup_update(struct sway_input_popup *popup) {
	struct sway_text_input *text_input =
		relay_get_focused_text_input(popup->relay);

	if (text_input == NULL || text_input->input->focused_surface == NULL) {
		return;
	}

	if (popup->scene_tree != NULL) {
		wlr_scene_node_destroy(&popup->scene_tree->node);
		popup->scene_tree = NULL;
	}
	if (popup->desc.relative != NULL) {
		wlr_scene_node_destroy(popup->desc.relative);
		popup->desc.relative = NULL;
	}

	if (!popup->popup_surface->surface->mapped) {
		return;
	}
```

Reply:
Ok, yes it should do in this way, thanks